matrix-synapse: 0.14 -> 0.16

Ralith opened the pull request:
This updates matrix-synapse to the less buggy 0.16 release, and updates and expands configuration support through the NixOS module. The python dependency bleach is also updated to the latest patch revision to satisfy matrix-synapse's requirements.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ralith committed:
bleach: 1.4 -> 1.4.3

Ralith committed:
matrix-synapse: 0.14 -> 0.16

roblabla commented:
Why remove bind_port/unsecure_bind_port ? It's still supported by synapse along with listeners. Having both would be really nice, as it'd avoid breaking existing installations of synapse :)

Also, you're mixing up spaces and tabs. It makes reading the nix file in my editor a bit painful.

roblabla commented on nixos/modules/services/misc/matrix-synapse.nix:
```diff
@@ -5,17 +5,22 @@ with lib;
   cfg = config.services.matrix-synapse;
   logConfigFile = pkgs.writeText "log_config.yaml" cfg.logConfig;
+  mkResource = r: ''{names: [${concatStringsSep "," r.names}], compress: ${if r.compress then "true" else "false"}}'';
+  mkListener = l: ''{port: ${l.port}, bind_address: ${l.bind_address}, type: ${l.type}, tls: ${if l.tls then "true" else "false"}, x_forwarded: ${if l.x_forwarded then "true" else "false"}, resources: [${concatStringsSep "," (map mkResource l.resources)}]}'';
```
port has to be a string here. That kinda sucks :( . It'd be nice if we could either do an if/else on the type and convert if necessary, or enforce port to be a number somehow.

joachifm replied:
It could be declared as a number and converted using the `toString` builtin.

roblabla replied:
if bind_address is empty, it also crashes matrix_synapse because `bind_address: ,` is translated to NoneType instead of an empty string (derp). It'd be a good idea to surround ${l.bind_address} with quotes.

Alternatively, why not use builtins.toJSON ? That should make everything work correctly and would be arguably "cleaner".

roblabla commented on nixos/modules/services/misc/matrix-synapse.nix:
```diff
@@ -94,46 +125,97 @@ in {
+      listeners = mkOption {
+        type = types.listOf types.attrs;
+        default = [{
+          port = 8448;
```
port has to be a string here, otherwise you get an error :

`cannot coerce an integer to a string, at <nixpkgs>/nixos/modules/services/misc/matrix-synapse.nix:10:19`

Ralith commented:
Updated with precise option types, restored backwards compatibility, and fixed assorted bugs. toJSON was not adopted because the precise option types cause various metadata to clutter up the objects.

Ralith commented:
Lots of iteration, but at this point I think I'm satisfied that this is ready to merge.

roblabla commented:
Built locally. It works :)
